File: erpnext/budget/report/supplementary_budget_report/supplementary_budget_report.py
# Copyright (c) 2022, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe import _
from frappe.utils import flt, getdate, formatdate, cstr

def execute(filters=None):
	validate_filters(filters)
	data = get_data(filters)
	columns = get_columns(filters)
	return columns, data

def validate_filters(filters):
	if not filters.fiscal_year:
		frappe.throw(_("Fiscal Year {0} is required").format(filters.fiscal_year))

	fiscal_year = frappe.db.get_value("Fiscal Year", filters.fiscal_year, ["year_start_date", "year_end_date"], as_dict=True)
	if not fiscal_year:
		frappe.throw(_("Fiscal Year {0} does not exist").format(filters.fiscal_year))
	else:
		filters.year_start_date = getdate(fiscal_year.year_start_date)
		filters.year_end_date = getdate(fiscal_year.year_end_date)

	if not filters.from_date:
		filters.from_date = filters.year_start_date

	if not filters.to_date:
		filters.to_date = filters.year_end_date

	filters.from_date = getdate(filters.from_date)
	filters.to_date = getdate(filters.to_date)

	if filters.from_date > filters.to_date:
		frappe.throw(_("From Date cannot be greater than To Date"))

	if (filters.from_date < filters.year_start_date) or (filters.from_date > filters.year_end_date):
		frappe.msgprint(_("From Date should be within the Fiscal Year. Assuming From Date = {0}")\
			.format(formatdate(filters.year_start_date)))

		filters.from_date = filters.year_start_date

	if (filters.to_date < filters.year_start_date) or (filters.to_date > filters.year_end_date):
		frappe.msgprint(_("To Date should be within the Fiscal Year. Assuming To Date = {0}")\
			.format(formatdate(filters.year_end_date)))
		filters.to_date = filters.year_end_date

def get_data(filters):
	query = """
		select cost_center, account, project, amount,
		(select a.remarks from `tabSupplementary Budget` a where a.name = `tabSupplementary Details`.reference) as remarks,
		posting_date as date 
		from `tabSupplementary Details` 
		where posting_date between '{from_date}' and '{to_date}'
		""".format(from_date=filters.from_date, to_date=filters.to_date)

	if filters.budget_against == "Project":
		if filters.to_project:
			query += " and project = \'" + filters.to_project  + "\'"
		else:
			query += " and (project !='' and project is NOT NULL)"
	else:
		query += " and (project = '' or project is NULL)"
		
	if filters.to_cc:
		query+=" and cost_center = \'" + filters.to_cc  + "\'"

	if filters.to_acc and filters.budget_against == "Cost Center":
		query+=" and account = \'" + filters.to_acc  + "\'"
	sup_data = frappe.db.sql(query, as_dict=True)

	data = []

	if sup_data:
		for a in sup_data:
			row = {
				"to_project": a.project,
				"to_cc": a.cost_center,
				"to_acc": a.account,
				"amount": a.amount,
				"date": a.date,
			}
			data.append(row)
	
	return data

def get_columns(filters):
	if filters.budget_against != "Project":
		return [
			{
				"fieldname": "date",
				"label": _("Date"),
				"fieldtype": "Date",
				"width": 120
			},
			{
				"fieldname": "to_cc",
				"label": _("To Cost Center"),
				"fieldtype": "Link",
				"options":"Cost Center",
				"width": 200
			},
			{
				"fieldname": "to_acc",
				"label": _("To Account"),
				"fieldtype": "Link",
				"options": "Account",
				"width": 230
			},
			{
				"fieldname": "amount",
				"label": _("Amount"),
				"fieldtype": "Currency",
				"width": 130
			},
			{
				"fieldname": "remarks",
				"label": _("Remarks"),
				"fieldtype": "Data",
				"width": 200
			}
		]
	else:
		return [
			{
				"fieldname": "date",
				"label": _("Date"),
				"fieldtype": "Date",
				"width": 120
			},
			{
				"fieldname": "to_project",
				"label": _("To Project"),
				"fieldtype": "Link",
				"options": "Project",
				"width": 230
			},
			{
				"fieldname": "to_cc",
				"label": _("To Cost Center"),
				"fieldtype": "Link",
				"options":"Cost Center",
				"width": 200
			},
			{
				"fieldname": "amount",
				"label": _("Amount"),
				"fieldtype": "Currency",
				"width": 130
			},
			{
				"fieldname": "remarks",
				"label": _("Remarks"),
				"fieldtype": "Data",
				"width": 200
			}
		]

File: erpnext/budget/report/supplementary_budget_report/supplementary_budget_report.js
// Copyright (c) 2022, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Supplementary Budget Report"] = {
	"filters": [
		{
			"fieldname": "fiscal_year",
			"label": __("Fiscal Year"),
			"fieldtype": "Link",
			"options": "Fiscal Year",
			"default": frappe.defaults.get_user_default("fiscal_year"),
			"reqd": 1,
			"on_change": function(query_report) {
				var fiscal_year = query_report.get_values().fiscal_year;
				if (!fiscal_year) {
					return;
				}
				frappe.model.with_doc("Fiscal Year", fiscal_year, function(r) {
					var fy = frappe.model.get_doc("Fiscal Year", fiscal_year);
					query_report.filters_by_name.from_date.set_input(fy.year_start_date);
					query_report.filters_by_name.to_date.set_input(fy.year_end_date);
					query_report.trigger_refresh();
				});
			}
		},
		{
			"fieldname": "from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"default": frappe.defaults.get_user_default("year_start_date"),
		},
		{
			"fieldname": "to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"default": frappe.defaults.get_user_default("year_end_date"),
		},
		{
			"fieldname":"budget_against",
			"label": __("Budget Against"),
			"fieldtype": "Select",
			"options": ["", __("Cost Center"), __("Project")],
			on_change: function(query_report){
				var budget_against = query_report.get_filter_value('budget_against');
				if(budget_against == "Project"){
					var to_acc = query_report.get_filter("to_acc"); to_acc.toggle(false);
					//to_acc.refresh();
					var to_project = query_report.get_filter("to_project"); to_project.toggle(true);
					//to_project.refresh();
				}else{
					var to_acc= query_report.get_filter("to_acc");
					to_acc.toggle(true);
					//to_acc.refresh();
					var to_project= query_report.get_filter("to_project");
					to_project.toggle(false);
					//to_project.refresh();
				}
				query_report.trigger_refresh();

			}
		},
		{
			"fieldname": "to_project",
			"label": __("To Project"),
			"fieldtype": "Link",
			"options": "Project",
		},
		{
			"fieldname": "to_cc",
			"label": __("To Cost Center"),
			"fieldtype": "Link",
			"options": "Cost Center",
		},
		{
			"fieldname": "to_acc",
			"label": __("To Account"),
			"fieldtype": "Link",
			"options": "Account",
		},
	]
}